core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java

```diff
@@ -0,0 +1,130 @@
+        try {
+            return path.toRealPath();
+        } catch (IOException e) {
+            throw new IllegalArgumentException("failed to locate file " + path, e);
+        }
+    }
+
+    private final Map<PropertiesEndpointGroup, FileWatcherContext> ctxRegistry =
+            new ConcurrentHashMap<>();
+    private final WatchService watchService;
+    private final RestartableThread restartableThread;
+
+    /**
+     * Create a registry using the default file system.
+     */
+    PropertiesFileWatcherRegistry() {
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        restartableThread =
+                new RestartableThread("armeria-file-watcher",
+                                      () -> new FileWatcherRunnable(watchService, ctxRegistry::values));
+    }
+
+    /**
+     * Registers a {@code path} of the properties file to be watched.
+     * @param filePath path to be watched
+     * @param reloader callback to be called on a file change event
+     */
+    synchronized void register(PropertiesEndpointGroup group, Path filePath, Runnable reloader) {
+        final Path dirPath = getRealPath(filePath).getParent();
+        final WatchKey key;
+        try {
+            key = dirPath.register(watchService,
+                                   ENTRY_CREATE,
+                                   ENTRY_MODIFY,
+                                   ENTRY_DELETE,
+                                   OVERFLOW);
+        } catch (IOException e) {
+            throw new IllegalArgumentException("failed to watch file " + filePath, e);
+        }
+        ctxRegistry.put(group, new FileWatcherContext(key, reloader, dirPath));
+        if (!ctxRegistry.isEmpty()) {
+            restartableThread.start();
+        }
+    }
+
+    /**
+     * Stops watching a properties file corresponding to the {@code PropertiesEndpointGroup}.
+     * @param group group to stop watching
+     */
+    synchronized void deregister(PropertiesEndpointGroup group) {
+        if (!ctxRegistry.containsKey(group)) {
+            return;
+        }
+        final FileWatcherContext removedCtx = ctxRegistry.remove(group);
+        final boolean existsDirWatcher = ctxRegistry.values().stream().anyMatch(
+                value -> value.path().equals(removedCtx.path()));
+        if (!existsDirWatcher) {
+            removedCtx.cancel();
+        }
+        if (ctxRegistry.isEmpty()) {
+            restartableThread.stop();
+        }
+    }
+
+    /**
+     * Check if future for {@code WatchService} is running.
+     * @return true if future is running
+     */
+    @VisibleForTesting
+    boolean isRunning() {
+        return restartableThread.isRunning();
+    }
+
+    /**
+     * Close watch service, future, and clear registry.
+     * @throws Exception may be thrown if {@code WatchService} is already closed
```

Comment:
From the javadoc for WatchService.close():
```
     * If this watch service is already closed then invoking this method
     * has no effect.
```
So this statement seems necessary to be checked again?

Reply:
good point:)